File: puzzles/hard/kotlin/src/roller-coaster.kt
import java.util.Scanner

fun main(args : Array<String>) {
  val input = Scanner(System.`in`)
  val nbPlaces = input.nextInt()
  val nbRides = input.nextInt()
  val nbGroups = input.nextInt()
  val groups = ArrayList<Long>()
  for (i in 0 until nbGroups) {
    val nbPeople = input.nextInt()
    groups.add(nbPeople.toLong())
  }

  // Calculate earnings starting from each group (1 ride)
  // Save the result (dynamic programming)
  val results = ArrayList<Pair<Long, Int>>()
  for (i in 0 until nbGroups) {
    var index = i
    var nbPeople = groups[index]
    var nbPeopleRiding = 0L
    while (nbPeopleRiding + nbPeople <= nbPlaces) {
      nbPeopleRiding += nbPeople
      if (index < nbGroups - 1) {
        index++
      } else {
        index = 0
      }
      // all groups are riding already?
      if (index == i) {
        break
      }
      nbPeople = groups[index]
    }
    results.add(Pair(nbPeopleRiding, index))
  }

  // Calculate the total earnings starting from the first group (all rides)
  var totalMoney = 0L
  var currentGroup = 0
  for (i in 0 until nbRides) {
    val (money, nextGroup) = results[currentGroup]
    totalMoney += money
    currentGroup = nextGroup
  }

  println(totalMoney)
}

File: puzzles/python3/roller-coaster/roller_coaster.py
# read game input
groups = []
nb_places, nb_rides, nb_groups = map(int, input().split())
for i in range(nb_groups):
    group_size = int(input())
    groups.append(group_size)

# cache the results starting to fill the ride from each of the groups
results = []
for i in range(nb_groups):
    group_index = i
    nb_places_left = nb_places
    while nb_places_left >= groups[group_index]:
        nb_places_left -= groups[group_index]
        group_index += 1
        if group_index == nb_groups:
            group_index = 0
        if group_index == i:
            break
    money = nb_places - nb_places_left
    results.append((money, group_index))

# calculate the total money earned during the day
total_money = 0
group_index = 0
for _ in range(nb_rides):
    money, next_group_index = results[group_index]
    total_money += money
    group_index = next_group_index

print(total_money)
